CL-1893: Add logic to surveys

EdwinKato opened the pull request:
All work to be released with the logic in surveys

## Checklist

- [x] Added entry to change log
- [ ] Reviewed by all members on the team

## Links

- [Ticket](https://citizenlab.atlassian.net/browse/CL-1893)

EdwinKato committed:
Hide WYSIWYGControl when visible is false

EdwinKato committed:
Fix useEffect dependency

EdwinKato committed:
Fix hasOwnProperty eslint error

EdwinKato committed:
Handle hidden elements on a page

EdwinKato committed:
Remove hasShowRule dead code

Commit:
[CL-2003] Merge branch 'CL-1892-pages' into CL-2003-logic-in-be-koen

Commit:
[CL-2106] WIP Add initial tab behaviour

Commit:
[CL-2106] WIP Add additional styling and functionality for settings tabs

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Add test for getPageSchema

EdwinKato committed:
Add test for isPageCategorization

EdwinKato committed:
Add unit tests for visibility

EdwinKato committed:
Clean up types and add more tests

EdwinKato committed:
Remove any cast

EdwinKato committed:
Add test to check that hidden fields are not passed in required key

Commit:
[CL-2106] WIP additional rule setup

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Remove hidden data from form submission and validation

EdwinKato committed:
Add test for getFormSchemaAndData

EdwinKato committed:
Add tests to verufy hidden fields are not pushed on submit

EdwinKato committed:
Remove debugger

Commit:
[CL-2106] WIP Add live data handling

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

Commit:
Translations updated by CI (extract-intl)

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2003-logic-in-be-koen # Conflicts: # back/db/schema.rb

koendehondt committed:
Merge pull request #3214 from CitizenLabDotCo/CL-2003-logic-in-be-koen

[CL-2003] logic in the BE

Commit:
[CL-2106] Add additional handling for logic in back office

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2106] Fix issue with new pages

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

EdwinKato committed:
Nest questions on page to create drag preview

Commit:
[CL-2106] Fix issue with live data sync

koendehondt committed:
[CL-2106] Simplify the expected logic payload by using goto_page_id for survey_end as well

Commit:
[CL-1893] Merge branch 'master' into CL-1893-logic

Commit:
[CL-2106] Merge branch 'CL-1893-logic' into CL-2106-add-logic-settings

Commit:
[CL-2106] Remove duplicate message

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2106] Ensure backwards compatibility

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

koendehondt committed:
[CL-2106] Make spec more robust

It failed on CI, but not locally. The failure indicated that the expected field order was not correct. The question came before the page.

Commit:
[CL-2106] Make temp ids more clear and add to new options

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

EdwinKato committed:
Handle dragging a page with questions

Commit:
[CL-2106] Code cleanup

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2106] Code cleanup

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

Commit:
[CL-2106] Only send tempID on field if it is a page

koendehondt committed:
[CL-2106] Handle temp ids for new options, and use ids instead of keys for option values in logic

koendehondt committed:
[CL-2106] Se do not support logic on multiselect fields

koendehondt committed:
[CL-2106] Update spec after switching from option keys to option ids in logic

Commit:
[CL-2106] Code cleanup

Commit:
[CL-2106] Merge branch 'CL-2106-add-logic-settings' of github.com:CitizenLabDotCo/citizenlab into CL-2106-add-logic-settings

EdwinKato committed:
Fix drag and drop for sticky first and last elements

EdwinKato committed:
Clean up code

Commit:
[CL-2106] Remove logic settings from multiselect

Commit:
[CL-2106] Only send logic for certain fields

Commit:
[CL-2106] Fix sending logic on save

EdwinKato committed:
Add missing key

EdwinKato committed:
Fix failing SortableRow test

EdwinKato committed:
Clean up code

EdwinKato committed:
Merge with CL-1893-logic

EdwinKato committed:
Merge pull request #3377 from CitizenLabDotCo/CL-2108-apply-logic-FO

CL-2108: Apply the logic when taking a survey

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2001-allow-dragging-and-dropping-pages-with-questions

Commit:
[CL-2106] Merge branch 'CL-1893-logic' into CL-2106-add-logic-settings

Commit:
[CL-2106] Remaining updates from code review

Commit:
[CL-2106] Fixes from code review

EdwinKato committed:
Check that all rules are fulfilled

EdwinKato committed:
Add test for all rules

EdwinKato committed:
Merge pull request #3432 from CitizenLabDotCo/bug/fixMultipleRulesBug

Bug/fix multiple rules bug

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2106-add-logic-settings

koendehondt committed:
Merge branch 'master' into CL-1893-logic

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2106-add-logic-settings

koendehondt committed:
[CL-2106] (code review) Ensure that multiple options with logic can be updated correctly

Commit:
[CL-2107] Add logic view to fields list

Commit:
[CL-2107] Update icon color

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2001-allow-dragging-and-dropping-pages-with-questions

Commit:
[CL-2107] Change icon positioning

EdwinKato committed:
Use react-beautiful-dnd instead

EdwinKato committed:
Clean up implementation

EdwinKato committed:
Fix drag behavior when page is empty

amanda-anderson committed:
Merge pull request #3419 from CitizenLabDotCo/CL-2106-add-logic-settings

[CL-2106] Add logic settings to form builder

EdwinKato committed:
Remove dead code

EdwinKato committed:
Remove unnecessary props

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2001-allow-dragging-and-dropping-pages-with-questions

Commit:
[CL-2107] Fix issue with rule display ordering

Commit:
[CL-2107] Merge with base branch

Commit:
[CL-2209] Remove icon from submit survey button

EdwinKato committed:
Clean up

koendehondt committed:
Merge branch 'master' into CL-1893-logic

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2107-add-logic-to-fields

EdwinKato committed:
Fix selected field after dragging

EdwinKato committed:
Fix alignment of page element

EdwinKato committed:
Add type for drag and drop result

Commit:
[CL-2209] Change survey page title to tenant primary colour

EdwinKato committed:
Add border during drag

EdwinKato committed:
Remove any type

EdwinKato committed:
Remove background

EdwinKato committed:
Merge pull request #3422 from CitizenLabDotCo/CL-2001-allow-dragging-and-dropping-pages-with-questions

CL-2001: Allow dragging and dropping pages with questions

Commit:
[CL-2209] Fix issues with locale selector

koendehondt committed:
Merge branch 'master' into CL-1893-logic

Commit:
[CL-2107] Merge with base branch

Commit:
[CL-2107] Merge branch 'CL-2107-add-logic-to-fields' of github.com:CitizenLabDotCo/citizenlab into CL-2107-add-logic-to-fields

Commit:
[CL-2107] Move functions to util file

Commit:
[CL-2107] Fix flickering row issue when dragging due to undefined locale

amanda-anderson committed:
Merge pull request #3434 from CitizenLabDotCo/CL-2107-add-logic-to-fields

[CL-2107] Add logic view to form builder field list

koendehondt commented on front/app/containers/Admin/formBuilder/components/FormBuilderSettings/index.tsx:
```diff
@@ -44,11 +45,39 @@ const FormBuilderSettings = ({
+  const getPageList = () => {
+    // TODO: Only select pages which come after the question. For this iteration though, we are not
+    // validating against cyclical form flows, so to not have an error state here,
+    // all pages should be available in the list.
```
When we implement the validation ticket, this TODO has to be removed or revised.

amanda-anderson replied:
> Are all translations available?

The translations were finished, but we'll be adding more now with the UX tweaks + FE validation. I'll put the label back on :)

koendehondt reviewed (commented):
Are all translations available?

Commit:
[CL-2209] Fix linear scale label settings style

EdwinKato committed:
Add tests to verify e2e flow for logic in pages

EdwinKato committed:
Fix linear scale test after change in logic to add end page

EdwinKato committed:
Fix number field test after change in logic to add end page

Commit:
[CL-2209] UI tweaks

EdwinKato committed:
Merge pull request #3474 from CitizenLabDotCo/CL-2239-add-update-e-2-e-tests-to-cover-logic

CL-2239: Add update e2e tests to cover logic

Commit:
[CL-2209] Use renamed variable in main app

Commit:
[CL-2209] Bump CL version and small UI tweaks

Commit:
[CL-2209] Merge with base branch

EdwinKato committed:
Add long text field

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Add e2e test for the long field

EdwinKato committed:
Correct describe name

Commit:
[CL-1861] UI tweaks

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Merge pull request #3477 from CitizenLabDotCo/CL-1861-ux-ui-tweaks-second-round

[CL-1861] Second round native survey UI tweaks

koendehondt committed:
[CL-2256] Support multiline text fields in the survey export to Excel

EdwinKato committed:
Merge pull request #3476 from CitizenLabDotCo/CL-2256-add-long-text-field-to-survey-builder

CL-2256: Add long text field

koendehondt committed:
[CL-1893] Improve spec description

EdwinKato committed:
Separate signle choice from multiple choice

EdwinKato committed:
Merge with CL-1893-logic

EdwinKato committed:
Update tests to match changes

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Update test description

EdwinKato committed:
Update with master and fix conflicts

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2285-separate-single-choice-field-from-multi-choice-in-form-builder

EdwinKato committed:
Remove unnecessary message entry

Commit:
[CL-2165] WIP Add logic validation and deletion confirmation modal

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Hide Rule Input when no option is added for single choice

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2165] WIP add logic validation to form builder

Commit:
[CL-2165] Merge branch 'CL-2165-add-validation-of-logic-to-the-survey-builder-fe' of github.com:CitizenLabDotCo/citizenlab into CL-2165-add-validation-of-logic-to-the-survey-builder-fe

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2165] WIP Fix validation for new fields

Commit:
[CL-2165] Merge branch 'CL-2165-add-validation-of-logic-to-the-survey-builder-fe' of github.com:CitizenLabDotCo/citizenlab into CL-2165-add-validation-of-logic-to-the-survey-builder-fe

Commit:
[CL-2165] Fix initial error state

Commit:
[CL-2165] Fix small bugs

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2165] Show logic error in canvas

Commit:
[CL-2165] Merge branch 'CL-2165-add-validation-of-logic-to-the-survey-builder-fe' of github.com:CitizenLabDotCo/citizenlab into CL-2165-add-validation-of-logic-to-the-survey-builder-fe

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2165] Small fixes

Commit:
[CL-2165] Merge branch 'CL-2165-add-validation-of-logic-to-the-survey-builder-fe' of github.com:CitizenLabDotCo/citizenlab into CL-2165-add-validation-of-logic-to-the-survey-builder-fe

EdwinKato committed:
Handle jumping to page in CLPageLayout

Commit:
[CL-2165] Rename variable

koendehondt committed:
[CL-2291] Support page logic in the payload and in the UI schema generation

Commit:
[CL-2165] Fix bug when deleting first page

amanda-anderson committed:
Update front/app/containers/Admin/formBuilder/components/FormBuilderSettings/index.tsx

Co-authored-by: Koen De Hondt <koen@dehondt.dev>

Commit:
[CL-2165] Rename delete function

Commit:
[CL-2165] Replace array map

Commit:
[CL-2209] Merge with base branch

Commit:
[CL-2209] Fix UI issue with admin comment component

EdwinKato committed:
Handle saving page with rules in back office

Commit:
[CL-2209] Fix spacing

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Merge pull request #3443 from CitizenLabDotCo/CL-2209-ui-tweaks

[CL-2209] In-platform survey UI tweaks

koendehondt committed:
[CL-2291] Do not put hide rules on the page with the corresponding logic

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2291-page-logic

koendehondt committed:
[CL-2291] Use real page ids in UI schema

Commit:
[CL-2165] Add unit test for yup validation

Commit:
[CL-2165] Catch error message

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2285-separate-single-choice-field-from-multi-choice-in-form-builder

EdwinKato committed:
Merge pull request #3484 from CitizenLabDotCo/CL-2285-separate-single-choice-field-from-multi-choice-in-form-builder

CL-2285: Separate single choice field from multi choice in form builder

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2291-page-logic

EdwinKato committed:
Show page logic in phone

koendehondt committed:
[CL-2291] Add hasRule to options of question in UI schema

EdwinKato committed:
Give precedence to rules on questions over page rules

EdwinKato committed:
Filter out hidden pages on save

EdwinKato committed:
Merge branch 'master' into CL-1893-logic

EdwinKato committed:
Merge branch 'CL-1893-logic' into CL-2291-page-logic

EdwinKato committed:
Fix type issues in FE build

koendehondt committed:
[CL-2291] Adapt test after introduction of hasRule in UI schema

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2165-add-validation-of-logic-to-the-survey-builder-fe

EdwinKato committed:
Fix logic object when adding a page

EdwinKato committed:
Fix build issue

EdwinKato committed:
Fix validation method

EdwinKato committed:
Merge pull request #3495 from CitizenLabDotCo/CL-2165-add-validation-of-logic-to-the-survey-builder-fe

[CL-2165] Add logic validation to form builder

EdwinKato committed:
Merge with CL-1893-logic

EdwinKato committed:
Add validation for page logic

EdwinKato committed:
Add error message to page rule

koendehondt committed:
[CL-2291] Remove unnecesary code

EdwinKato committed:
Fix trigger validation on selecting another page

koendehondt committed:
[CL-2291] Question-level logic trumps page-level logic

When page-level logic is present, apply it for answers that are not invoved in question-level logic.

EdwinKato committed:
Handle deletion of page with associated rules

EdwinKato committed:
Code clean-up

koendehondt committed:
[CL-2291] Question-level logic trumps page-level logic for linear scale fields

koendehondt committed:
Merge branch 'CL-2291-page-logic' of https://github.com/CitizenLabDotCo/citizenlab into CL-2291-page-logic

EdwinKato committed:
Add e2e test for showing errors in logic

koendehondt committed:
[CL-2291] Allow { 'rules' => [] } as empty logic

koendehondt committed:
[CL-2291] Extract CustomField#logic? to remove code duplication

EdwinKato committed:
Use consistent badge for questions

EdwinKato committed:
Change page rule label

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Bump up cl2-library version

EdwinKato committed:
Change to use updated survey builder icons

EdwinKato committed:
Update style for required badge

EdwinKato committed:
Change to show required badge before field badge

EdwinKato committed:
Add icons to question badge

EdwinKato committed:
Change page rule display styles

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Remove canvas padding

EdwinKato committed:
Add row hover styles

EdwinKato committed:
Change text color when selected

EdwinKato committed:
Handle hover and error state colors

EdwinKato committed:
Add survey support article

EdwinKato committed:
Make questions with logic required

Commit:
Translations updated by CI (extract-intl)

EdwinKato committed:
Fix logic warning message

EdwinKato committed:
Rename RuleInput to QuestionRuleInput

EdwinKato committed:
Use consistent types for question rules

EdwinKato committed:
Code clean-up

EdwinKato committed:
Make adding page logic consistent with question logic

EdwinKato committed:
Hanlde long answer text

EdwinKato committed:
Improve hover and slected colors

EdwinKato committed:
Ensure that badge doesn't wrap

koendehondt committed:
[CL-2291] Return a validation error when a field with logic is not required

EdwinKato committed:
Add test to verify logic on the same page

EdwinKato committed:
Add test for page logic

EdwinKato committed:
Add e2e test for question logic prioritisation over page logic

koendehondt committed:
Merge branch 'master' into CL-1893-logic

koendehondt committed:
Merge branch 'CL-1893-logic' into CL-2291-page-logic

koendehondt committed:
[CL-2291] Start adding file_upload field to BE

koendehondt committed:
Revert "[CL-2291] Start adding file_upload field to BE" This reverts commit e9d784a6e2ce6929b6d48a1dfa77a5632e0780d2.

Commit:
[CL-2291] Rename message key

Commit:
Translations updated by CI (extract-intl)

Commit:
[CL-2291] Fix page logic deletion

Commit:
[CL-2291] Merge branch 'CL-2291-page-logic' of github.com:CitizenLabDotCo/citizenlab into CL-2291-page-logic

koendehondt committed:
Merge pull request #3496 from CitizenLabDotCo/CL-2291-page-logic

[CL-2291] Handle jumping to page in CLPageLayout

Commit:
[CL-1893] Merge with master

Commit:
[CL-1893] Merge branch 'CL-1893-logic' of github.com:CitizenLabDotCo/citizenlab into CL-1893-logic

Commit:
[CL-1893] Fix e2e after merge

koendehondt committed:
Merge branch 'master' into CL-1893-logic

Commit:
Filter data before submission

Commit:
Rename function for clarity

Commit:
Fix small bug

Commit:
Remove unnecessary check

Commit:
Improvements from code review and unit test

Commit:
Update e2e tests

amanda-anderson committed:
Merge pull request #3535 from CitizenLabDotCo/handle-form-submission-data-pages

Filter survey data before submission

Commit:
[CL-1893] Update changelog

Commit:
[CL-1893] Merge branch 'CL-1893-logic' of github.com:CitizenLabDotCo/citizenlab into CL-1893-logic

koendehondt reviewed (approved):
Lots of tests 👍
